### PR TITLE
Integrate with `tmp-postgres`

### PR DESCRIPTION
**Pull request by parsonsmatt**
This PR integrates the test suite with the fantastic [`tmp-postgres`](https://hackage.haskell.org/package/tmp-postgres) to ease the testing story.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -29,6 +29,7 @@ before_install:
 # Download and unpack the stack executable
 - mkdir -p ~/.local/bin
 - export PATH=$HOME/.local/bin:$PATH
+- export PATH=$PATH:/usr/lib/postgresql/10/bin/
```
Comment (parsonsmatt) [inline .travis.yml:32]: We need this in order for the `tmp-postgres` to find `initdb`. Ubuntu doesn't have `initdb` on the path by default.

File: schema/Usage.persistentmodels [hunk 2]
```diff
@@ -1,3 +1,5 @@
 Usage
     ride        RideId
     component   ComponentId
+
+    deriving Show Eq
```
Comment (parsonsmatt) [inline schema/Usage.persistentmodels:5]: t e s t i n g

File: src/GT/DB/Usage.hs [hunk 3]
```diff
@@ -21,7 +21,12 @@ import GT.DB.Schema.Ride
 -- TODO: esqueleto does not support CTEs. It should!
 new :: (MonadIO m) => BikeId -> RideId -> SqlPersistT m [UsageId]
 new bikeId rideId = do
-    components <- [sqlQQ|
+    mride <- get rideId
+    case mride of
+        Nothing ->
+            pure []
+        Just _ -> do
+            components <- [sqlQQ|
```
Comment (parsonsmatt) [inline src/GT/DB/Usage.hs]: Optimistically only doing the potentially expensive query after the ride is not existing. Probably unnecessary.

File: test/Spec/GT/DB/Usage.hs [hunk 4]
```diff
@@ -1,45 +1,47 @@
...
-                            , componentModel = "Asym 29"
-                            , componentDescription = "it is p good"
-                            , componentParent = Just componentId
-                            }
-                        rideId <- insert Ride
-                            { rideDistance = 32
-                            , rideDate = UTCTime (fromGregorian 2020 01 03) 0
-                            }
+        it "works" $ \conn -> do
+            usages <- runTestDb conn $ do
```
Comment (parsonsmatt) [inline test/Spec/GT/DB/Usage.hs]: So the pattern here is:

```haskell
it "does a thing" $ \conn -> do
    value <- runTestDb conn queryStuff
    makeAssertionsOn value
```

File: test/Spec/GT/DB/Prelude.hs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+
+-- | Provide a shared database connection to the tests. This summons up
+-- a database connection, runs migrations, and passes it to the test
+-- suites scoped under it.
+provideDatabase :: SpecWith TestDb -> Spec
+provideDatabase = do
+    aroundAll $ \action -> do
+        eresult <- Temp.with $ \db ->
+            runNoLoggingT $
+            withPostgresqlConn (Temp.toConnectionString db) $ \conn -> do
```
Comment (parsonsmatt) [inline test/Spec/GT/DB/Prelude.hs:82]: `tmp-postgres` exposes `toConnectionString` which makes it crazy easy to connect to a database for persistent. Way easier than I thought it would be.

File: src/GT/DB/Usage.hs [hunk 6]
```diff
@@ -7,46 +7,29 @@ module GT.DB.Usage
...
+import           GT.DB.Schema.Component
+import           GT.DB.Schema.Ride
+import           GT.DB.Schema.Usage
 
 -- | Create a new 'Usage' for all components on the 'Bike' for the given
 -- 'Ride'.
 --
 -- TODO: esqueleto does not support CTEs. It should!
-new :: (MonadIO m) => BikeId -> RideId -> SqlPersistT m [UsageId]
+new :: (MonadIO m) => BikeId -> RideId -> SqlPersistT m [Entity Usage]
```
Comment (parsonsmatt) [inline src/GT/DB/Usage.hs:23]: remember to write functions that are amenable to testing!

File: src/GT/DB/Component.hs [hunk 7]
```diff
@@ -21,3 +23,22 @@ mileage componentId =
...
+
+    UNION
+
+    SELECT c_r.*
+    FROM component AS c_r
+    INNER JOIN children AS c
+        ON c.id = c_r.parent
+    )
+SELECT ??
+FROM children AS component
```
Comment (parsonsmatt) [inline src/GT/DB/Component.hs:43]: This query was factored out so I could write a better test for it. The test for `usage` started to fail when I made it better!

File: test/Spec/GT/DB/Component.hs [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+                    , componentBrand = "Salsa"
+                    , componentModel = "Rustler"
+                    , componentDescription = ""
+                    , componentParent = Just frame2Id
+                    }
+
+                components <- Component.getForBike bikeId
+                map (componentName . entityVal) components
+                    `shouldMatchList`
+                        map componentName [ wheel, tire, handlebars, frame ]
```
Comment (parsonsmatt) [inline test/Spec/GT/DB/Component.hs]: This integration test is Good, because it has a few properties:

1. Checks that recursively known elements are present (eg `tires` are attached to `wheel`)
2. The order of the list doesn't matter, so the database can switch ordering and the test isn't messed up.
3. The error presented is *readable* - the `componentName` is much easier to parse than the rest of the stuff!

File: test/Spec/GT/DB/Prelude.hs [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+-- | Run the given action and rollback the transaction that created it.
+-- This allows you to insert whatever records you need into the database
+-- while not having to worry about cleaning them up.
+runTestDb :: TestDb -> SqlPersistT IO a -> IO a
+runTestDb (TestDb conn) action =
+    flip runSqlConn conn $ do
+        result <- action
+        transactionUndo
+        pure result
```
Comment (parsonsmatt) [inline test/Spec/GT/DB/Prelude.hs:68]: This is what makes the database testing fast. We don't worry about truncating tables or deleting old state, we just roll back the transaction.

File: test/Spec/GT/DB/Prelude.hs [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+                flip runSqlConn conn $ do
+                    runMigrationSilent (const id makeUnlogged migrateAll)
+                liftIO $ action $ TestDb conn
+        case eresult of
+            Left e -> throwM e
+            Right _ -> pure ()
+
+makeUnlogged :: Migration -> Migration
+makeUnlogged =
+    mapWriterT (censor (fmap (fmap (Text.replace "CREATe TABLE" "CREATe UNLOGGED TABLE"))))
```
Comment (parsonsmatt) [inline test/Spec/GT/DB/Prelude.hs:92]: `UNLOGGED` tables are significantly faster than logged tables for big inserts. Since we'll presumably be doing lots of inserts in scaffolding the tests, we want them unlogged. This isn't good for durabiliy, but it's a test suite, who cares?

File: test/Spec/GT/DB/Prelude.hs [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+-- a database connection, runs migrations, and passes it to the test
+-- suites scoped under it.
+provideDatabase :: SpecWith TestDb -> Spec
+provideDatabase = do
+    aroundAll $ \action -> do
+        eresult <- Temp.with $ \db ->
+            runNoLoggingT $
+            withPostgresqlConn (Temp.toConnectionString db) $ \conn -> do
+                flip runSqlConn conn $ do
+                    runMigrationSilent (makeUnlogged migrateAll)
```
Comment (parsonsmatt) [inline test/Spec/GT/DB/Prelude.hs:84]: Presumably, in this block, we'll eventually want to allow the user to provide a *seed* action that will be shared among all tests.

File: src/GT/DB/Usage.hs [hunk 12]
```diff
@@ -7,46 +7,29 @@ module GT.DB.Usage
...
-    FROM component AS c_r
-    INNER JOIN children AS c
-        ON c_r.parent = c.id
-    )
-SELECT ??
-FROM bike AS b
-INNER JOIN children AS component
-    ON component.parent = b.component
-WHERE b.component = #{bikeId}
-        |]
```
Comment (parsonsmatt) [inline src/GT/DB/Usage.hs:41]: Functions shouldn't be too complex. This function was actually wrong, because this query was incorrect. The query was factored out into it's own function, tested, and corrected. Then this function was correct again.

File: test/Spec/GT/DB/Component.hs [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+import Database.Persist.Postgresql
+
+import GT.DB.Schema.Component
+import GT.DB.Schema.Bike
+import qualified GT.DB.Component as Component
+
+spec :: SpecWith TestDb
+spec = do
+    describe "getForBike" $ do
+        itDb "works" $ do
```
Comment (parsonsmatt) [inline test/Spec/GT/DB/Component.hs:14]: Make things easy for users. The pattern `it msg $ \conn -> runTestDb conn action` will happen frequently with database tests, so you might as well make it easy.

File: test/Spec/GT/DB/Component.hs [hunk 14]
```diff
@@ -0,0 +1,74 @@
...
+                , componentBrand = "Salsa"
+                , componentModel = "Rustler"
+                , componentDescription = ""
+                , componentParent = Just frame2Id
+                }
+
+            components <- Component.getForBike bikeId
+            map (componentName . entityVal) components
+                `shouldMatchList`
+                    map componentName [ wheel, tire, handlebars, frame ]
```
Comment (parsonsmatt) [inline test/Spec/GT/DB/Component.hs:74]: A good test should directly measure what you care about while ignoring stuff you don't care about. Previously, this was ``length components `shouldBe` 2``, which succeeded with the wrong answer! Then the test silently failed because `components` became `Either a [x]`. But then I made this very explicit test, which captured a bug, and made it easy to debug and fix.

File: test/Spec/GT/DB/Usage.hs [hunk 15]
```diff
@@ -1,45 +1,58 @@
...
+                    , componentParent = Just wheelId
+                    }
+            tireId <- insert tire
+            rideId <- insert Ride
+                { rideDistance = 32
+                , rideDate = UTCTime (fromGregorian 2020 01 03) 0
+                }
 
-                        Usage.new bikeId rideId
-            length usageIds `shouldBe` 2
```
Comment (parsonsmatt) [inline test/Spec/GT/DB/Usage.hs:45]: This is such a bad test! It was succeeding correctly with `[UsageKey 1, UsageKey 1]`, because we're not measuring what we actually care about.

File: test/Spec/GT/DB/Usage.hs [hunk 16]
```diff
@@ -1,45 +1,58 @@
...
+            usages <- Usage.new bikeId rideId
+            map entityVal usages
+                `shouldMatchList`
```
Comment (parsonsmatt) [inline test/Spec/GT/DB/Usage.hs:54]: `shouldMatchList` is great because you don't need to care about list ordering.

File: test/Spec/GT/Prelude.hs [hunk 17]
```diff
@@ -1,8 +1,23 @@
...
+    , shouldNotBe
+    , shouldNotContain
+    , shouldNotReturn
+    , shouldNotSatisfy
+    , shouldReturn
+    , shouldSatisfy
+    , shouldStartWith
+    , expectationFailure
+    )
+import Test.Hspec.Expectations.Lifted
```
Comment (parsonsmatt) [inline test/Spec/GT/Prelude.hs:23]: Lifted expectations are nice, because they can be used directly in the `SqlPersistT` monad.

**Review by parsonsmatt — COMMENTED** [review]
Ready to go! Database testing in an easy and convenient manner.

File: package.yaml [hunk 18]
```diff
@@ -22,27 +22,36 @@ description:         Please see the README on GitHub at <https://github.com/gith
 
 dependencies:
```
Comment (friedbrice) [inline package.yaml:23]: In practice, can we add these as test dependencies, or do we need them in the app?

File: src/GT/DB/Component.hs (hunk 7 above)
Comment (codedmart) [reply src/GT/DB/Component.hs:43]: Is the `factored` referring to the use of raw sql here?

File: src/GT/DB/Usage.hs (hunk 12 above)
Comment (friedbrice) [reply src/GT/DB/Usage.hs:41]: I suppose the function was made testable by factoring out the instructions as a data structure and then inspecting that data structure, decoupling the description from the execution. However, the part that was factored out is abstract, as a `MonadIO m => SqlPersistT m (Entity _))`. That doesn't really feel like a concrete data structure that's amenable to inspection, so how did you go about the testing?

File: src/GT/DB/Component.hs (hunk 7 above)
Comment (codedmart) [reply src/GT/DB/Component.hs:43]: More or less I was interested in why you are using a sql quasi quoter here.

File: test/Spec/GT/DB/Prelude.hs (hunk 9 above)
Comment (friedbrice) [reply test/Spec/GT/DB/Prelude.hs:68]: nice

File: test/Spec/GT/DB/Component.hs (hunk 14 above)
Comment (friedbrice) [reply test/Spec/GT/DB/Component.hs:74]: The scaffolding is all nice, and I'm happy to see it worked out in detail here. I guess the hard part for me, then, is knowing what's worthwhile to test. Can you explain this test a bit, and the reasoning that led you to want to test this function and the reasoning that led you to how to test it this way?

File: src/GT/DB/Component.hs (hunk 7 above)
Comment (parsonsmatt) [reply src/GT/DB/Component.hs:43]: esqueleto does not support common table expressions or `UNION` so I needed to drop to raw SQL here.

Beforehand, the function `Usage.new` looked like:

```haskell
new bikeId rideId = do
  componentIds <- [rawQQ|
the
whole
big
query
  |]
  for componentIds $ \componentId -> do
    insert $ Usage bikeId componentId
```

In the test I wrote for `Usage.new`, i asserted that ``length usages `shouldBe` 2``, and this was indeed the answer I got back. However, that was because the underlying `componentIds` query was incorrect. I separated the SQL query into it's own function, and had it return `component`s instead, and then I could write a good test that verified that it did what I wanted it to.

This function then becomes a somewhat trivial composition of functions and doesn't really need a test, but I kept it anyway.

File: test/Spec/GT/DB/Component.hs (hunk 14 above)
Comment (friedbrice) [reply test/Spec/GT/DB/Component.hs:74]: E.g. for me it's obvious that if I make an obtuse but fast version of an obvious but slow algorithm, I want to write a test that checks that the two implementations agree (at least on a decent subset of the domain). That's clear. But how do you decide what database tests to write, and how do you decide how to write them?

File: src/GT/DB/Component.hs (hunk 7 above)
Comment (codedmart) [reply src/GT/DB/Component.hs:43]: Ah right, thanks!

**Review by friedbrice — COMMENTED** [review]
Wonderful scaffolding. It's nice to have a complete example laid out that I can refer to. I have some questions on the practical philosophy, but they're perhaps beyond the scope of this demonstration, so feel free to ignore them if they are.

File: src/GT/DB/Usage.hs (hunk 12 above)
Comment (parsonsmatt) [reply src/GT/DB/Usage.hs:41]: This function was made more testable by returning more information about what it did. Beforehand, it was returning `[UsageId]`, which is essentially totally opaque - an `Id` is essentially a totally random value returned by the database signifying that it inserted something correctly. The previous test expected it to insert two `Usage` records, so I checked ``length results `shouldBe` 2``, which passed. 

By returning the entity `Entity Usage`, I could write a more specific test - that the two `Usage`s should associate the `Ride` and both `Component`s. This discovered that the two `Usage`s that were being returned referred to the same component! Oops.

Comment (parsonsmatt) [reply src/GT/DB/Usage.hs:41]: For functionality that depends entirely on the database's behavior, we want to verify our assumptions and encode our expectations.

File: package.yaml (hunk 18 above)
Comment (parsonsmatt) [reply package.yaml:23]: Yeah, you can have them as test-only dependencies.

File: src/GT/DB/Component.hs (hunk 7 above)
Comment (parsonsmatt) [reply src/GT/DB/Component.hs:43]: Beforehand, the query lived in `Usage.new`. I factored it out to `Component.getForBike`, so that I could test it independently of `Usage.new`.

I'm using a raw SQL here because esqueleto doesn't support UNION or recursive CTEs.

File: src/GT/DB/Usage.hs (hunk 6 above)
Comment (maddie927) [reply src/GT/DB/Usage.hs:23]: Out of curiosity, what makes this version more testable?

Comment (parsonsmatt) [reply src/GT/DB/Usage.hs:23]: Github is hiding the conversation that covers it, but this is a better place to answer it anyway:

https://github.com/parsonsmatt/gear-tracker/pull/13#discussion_r370277080

tl;dr: `UsageId` is an opaque identifier that doesn't tell you anything about the `Usage`. `Entity Usage` has the same information but also tells you about *what* the `Usage` is. This lets you write tests against the actual values that are being inserted.

File: test/Spec/GT/DB/Component.hs (hunk 14 above)
Comment (parsonsmatt) [reply test/Spec/GT/DB/Component.hs:74]: >  Can you explain this test a bit, and the reasoning that led you to want to test this function and the reasoning that led you to how to test it this way?

Yeah, definitely!

The function is defined as a blob of raw SQL because the database library does not support features that are required to express it. So I have a few things I want to verify right off the bat:

1. I wrote syntactically correct SQL (need Postgres to report on this)
2. I wrote SQL that the database library can handle (eg the `??` placeholders need to expand correctly, and I need to ensure I have the relevant columns prepared)

Basically, I need a test that just *runs* the query and doesn't do anything else to prevent somewhat predictable run time errors. I'd say this is testing the 'form' of the code, in the same way that you need GHC to tell you if your Haskell code is well-formed. We need to have `persistent-qq` and `postgres` verify this is well-formed.

After that, there's the *behavior* of the code. I write code and I have some expectation on how it works. I expect this function to return all of the components that belong to a bike, including components that belong to components that belong to the bike, recursively, etc. It's my first recursive CTE ever and I have no idea if I'm doing it right, so testing is a good way to explore this. Playing in the REPL is a good way to explore it, too, but this is automated. I expect the function to also not return any components that are on *other* bikes, or not on the bike, etc. 

TO an experienced DBA, the SQL is probably *obviously* correct. I am not an experienced DBA. But i also want to ensure that *calling* the code is convenient and fine, too! So I want to write a test that captures the things i care about.

This test is Not Ideal because it tests *all* of the things I describe above - if any of them break, the whole test breaks, and it's my job to inspect the test output to figure out *which* property broke. Indeed, a better approach would look like:

```haskell
describe "getForBike" $ before performInserts $ do
    it "is well-formed SQL" (shouldNotThrowException (runQuery getForBike bikeId rideId))
    it "has all components for a bike" (runQuery ... assert has all components)
    it "has no components for another bike" $ ...
    it "has recursive components for a bike" $ ...
```

Anyway, I structure the test to capture all those properties. So first I need the database to have a given state. I need it to contain two bikes, and a bunch of components. Most of those components should belong to the first bike (which I am running the test on), and some of the components should belong to *components* on the first bike (to test recursive part of query).

Then I run the query, and I can make assertions against the output I get. In this case, I just say that "All of the component's returned must be what I inserted."
